### PR TITLE
Enable fused foreach Adam compilation

### PR DESCRIPTION
**Pull request by mlazos**
Fixes #ISSUE_NUMBER

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @anijain2305

### CODE REVIEW

**Comment by mlazos** [comment]
Currently getting NaNs so I'm investigating that, but wanted to get feedback on the optim changes

File: torch/optim/adam.py [hunk 1]
```diff
@@ -99,7 +100,7 @@ def _init_group(
                     # This is because kernel launches are costly on CUDA and XLA.
                     state['step'] = (
                         torch.zeros((), dtype=torch.float, device=p.device)
-                        if group['capturable'] or group['fused']
+                        if group['capturable'] or group['fused'] or is_compiling
```
Comment (janeyx99) [inline torch/optim/adam.py:103]: Could you explain again by this cannot be the following?  Adding another argument in the _init_group and making adding is_compiling to this code is very unideal.
```suggestion
                        if group['capturable'] or group['fused'] or torch._utils.is_compiling()
```

Comment (janeyx99) [reply torch/optim/adam.py:103]: Also, this will run into a similar problem for the special case of https://github.com/pytorch/pytorch/issues/103256. We can avoid this problem by asking the user to turn capturable=True when using PT2 and adding a check to verify that capturable=True when using torch.compile. We can discuss this in more detail on Tuesday.

Comment (mlazos) [reply torch/optim/adam.py:103]: Hmm yeah it probably will. I'd prefer not to have a user have to set a flag, since this should be fully automatic. I could have dynamo set a hidden flag on the optimizer which is checked here, and also checked during serde if that's acceptable?

Comment (mlazos) [reply torch/optim/adam.py]: FYI removed this and added an internal flag which we can use as we see fit

**Comment by mlazos** [comment]
> Currently getting NaNs so I'm investigating that, but wanted to get feedback on the optim changes

Fixed by https://github.com/pytorch/pytorch/pull/104137

**Review by janeyx99 — COMMENTED** [review]
Instead of using self.compiled and introducing a new "internal-only" flag, which...meh, not good engineering, we can just overload self.capturable. So dynamo can just set capturable to True when compiling.

Then, we're guaranteed that step will follow to the right device. e.g., rn one would need to update line 432 in optimizer.py to check the flag so step could be loaded correctly.

File: torch/optim/optimizer.py
Comment (janeyx99) [inline torch/optim/optimizer.py:1]: Could you also update the comment in line 255 onwards about "# If we are compiling, we take the capturable path automatically" now that we do change the capturable flag? We may also be able to simplify some of the code there.

File: torch/optim/optimizer.py [hunk 2]
```diff
@@ -422,11 +422,16 @@ def _process_value_according_to_param_policy(param: Tensor, value: Tensor, param
                 capturable = pg["capturable"] if "capturable" in pg else False
                 break
 
-        if key != "step" or capturable or fused:
+        if key == 'step':
+            if capturable or fused:
+                return value.to(dtype=torch.float32, device=param.device)
+            else:
+                return value
+        else:
```
Comment (janeyx99) [inline torch/optim/optimizer.py:430]: I'd rewrite this for loop this way:
```
if key == 'step':
    if capturable or fused:
        return value.to(dtype=torch.float32, device=param.device)
    # otherwise, we DON'T cast step in any way.
    return value
if param.is_floating_point():
    return value.to(dtype=param.dtype, device=param.device)
return value.to(device = param.device)
```

This also means we're saying "don't touch step UNLESS it should join the parameter for fused/capturable". I'm curious if we should cast it to float in all cases.

File: torch/optim/adamw.py [hunk 3]
```diff
@@ -479,7 +479,7 @@ def _multi_tensor_adamw(
     if len(params) == 0:
         return
 
-    if capturable:
+    if not torch._utils.is_compiling() and capturable:
```
Comment (janeyx99) [inline torch/optim/adamw.py:482]: Will inductor auto move params and state_steps to be CUDA tensors?

File: torch/optim/adamw.py [hunk 4]
```diff
@@ -304,7 +304,7 @@ def adamw(
     See :class:`~torch.optim.AdamW` for details.
     """
 
-    if not all(isinstance(t, torch.Tensor) for t in state_steps):
+    if not torch._utils.is_compiling() and not all(isinstance(t, torch.Tensor) for t in state_steps):
```
Comment (janeyx99) [inline torch/optim/adamw.py:307]: Wait I'm confused why this one is added--wouldn't state_steps still be full of singleton tensors?

File: torch/optim/adam.py [hunk 5]
```diff
@@ -428,7 +428,9 @@ def _multi_tensor_adam(params: List[Tensor],
     if len(params) == 0:
         return
 
-    if capturable:
+    # If we are compiling, the compiler will determine if cudagraphs
```
Comment (janeyx99) [inline torch/optim/adam.py:431]: I appreciate the comment here--I wonder if there's a more discoverable place for it. You could write this once (in the first instance), and then name it as a NOTE: [compilers handle CUDAgraphs] and reference that everywhere.

I'd also appreciate knowing what the compiler would do + what checks we don't expect it to pass anymore.

File: torch/optim/adamw.py (hunk 3 above)
Comment (mlazos) [reply torch/optim/adamw.py:482]: no but it will partition the graph if they're not, so there shouldn't be any issue here

File: torch/optim/adamw.py (hunk 4 above)
Comment (mlazos) [reply torch/optim/adamw.py:307]: Yeah this assert is slow in dynamo. I can check this internally if you prefer, but I honestly felt like this assert isn't needed as much anymore anyway? In the usual case, the optimizer itself sets these tensors up, this is only if the user is directly setting up the optimizer states and calling this method manually.

File: torch/optim/optimizer.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/optimizer.py:430]: I feel like omitting the else blocks makes this code less clear tbh, if this is a nit, imma not fix. Let me know if you feel really strongly about this.

Comment (mlazos) [reply torch/optim/optimizer.py:430]: The way the initialization is written the step ignores the dtype of the param, it's either fp64  by default or fp32 if capturable/fused, so this matches that behavior.

File: torch/optim/adam.py (hunk 5 above)
Comment (mlazos) [reply torch/optim/adam.py:431]: I just copied it since it's only used in two places and it's pretty short

File: torch/optim/optimizer.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/optimizer.py:430]: I think the comments are more useful than the else blocks (they just make the indentation deeper).

Comment (janeyx99) [reply torch/optim/optimizer.py:430]: Yea lol I'm aware of the way it's been written, I'm more intrigued by whether we should have casted to fp32 vs fp64, but your change is fine for now. My question is beyond the scope and can be rectified later if need be.

File: test/inductor/test_compiled_optimizers.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+            def fn():
+                step_fn(opt_compiled, closure)
+
+        else:
+
+            def fn():
+                step_fn(opt_compiled)
+
+        with torch.set_grad_enabled(False):
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:63]: question: does set_grad_enabled default to True even in a test case? / why is this call necessary for the torch.compile call?

Comment (janeyx99) [reply test/inductor/test_compiled_optimizers.py:63]: oh is it for the step, cuz usually we disable grad for that? but since we ignore the wrapper, we have to replicate the logic to disable grad?

File: torch/optim/adamw.py (hunk 4 above)
Comment (janeyx99) [reply torch/optim/adamw.py:307]: wait so is the only reason this is skipped cuz it’s slow? it doesn’t break any tests?

this check was introduced to ensure steps are tensors and no longer python scalars cc @mikaylagawarecki to confirm—will torch.compile ensure this somewhere down the line? as tracing also requires step be tensors right?

if the sole reason is cuz tracing this is slow, that’s a good thing to point out in a comment

File: torch/optim/adam.py (hunk 5 above)
Comment (janeyx99) [reply torch/optim/adam.py:433]: alright i would appreciate the comment more if it addressed why the check no longer works. my previous understanding was that it’s due to step not necessarily being on cuda when torch.compiled—is this correct?

will inductor handle moving step later? i’d rather have the comment be specific here than the vague “just let the compiler worry about it.” also, i had previously thought of the torch compiled version as a capturable implementation which is why the conflation is okay. in other words, torch compiling an optim will give u the best capturable result. will we run into issues trying to differentiate capturable when pt2 vs capturable in eager?

File: test/inductor/test_compiled_optimizers.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    def setUp(self):
+        super().setUp()
+        torch._inductor.metrics.reset()
+
+    def tearDown(self):
+        super().tearDown()
+        torch._inductor.metrics.reset()
+
+    test_adam = make_test(torch.optim.Adam, lr=0.01, foreach=True)
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:91]: the foreach=True should be default, so would be a better test to not specify it to ensure the default is traceable and compilable. or test both. 

shall we try some other configs? especially weight decay?

**Review by janeyx99 — COMMENTED** [review]
Did a deeper dive of the changes. I think we should clarify what users should expect when they torch compile vs capturable vs both. I had previously thought of torch.compile be an overriding of capturable…but is there a space where the eager capturable will vary from torch compiled step?

File: test/inductor/test_compiled_optimizers.py (hunk 6 above)
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:63]: Yeah you got it

File: torch/optim/adamw.py (hunk 4 above)
Comment (mikaylagawarecki) [reply torch/optim/adamw.py:307]: @janeyx99 yep that's right the reasoning for adding this check at that point was if I recall correctly 

- these functions were moved from [`torch/optim/_functional.py`](https://github.com/pytorch/pytorch/blob/2a5aaf1c49f18e91c911d6020b71a4201c734eb4/torch/optim/_functional.py) and were technically private APIs that we don't offer a contract on
- these private APIs might be used by internal so we wanted to fail instead of causing silent breakages

Comment (mlazos) [reply torch/optim/adamw.py:307]: Yeah the only reason I skipped this is because it's slow. I can add a comment

File: test/inductor/test_compiled_optimizers.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+from torch.testing._internal.common_utils import TestCase
+
+
+aten = torch.ops.aten
+
+try:
+    try:
+        from .test_torchinductor import check_model, check_model_cuda, requires_cuda
+    except ImportError:
+        from test_torchinductor import check_model, check_model_cuda, requires_cuda
```
[the remark this answers is not in the document]
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:21]: Not really sure, but the other torchinductor test suites do this as well, so I think it was compatibility with different environments/versions or something?

File: test/inductor/test_compiled_optimizers.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        sys.exit(0)
+    raise
+
+
+def make_test(optim_cls, closure=None, **kwargs):
+    @requires_cuda()
+    def test_fn(self):
+        input = torch.ones([10, 10], device="cuda:0")
+        model_eager = torch.nn.Sequential(
+            *[torch.nn.Linear(10, 10, device="cuda:0") for _ in range(2)]
```
[the remark this answers is not in the document]
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:34]: idk, maybe just habit haha, this is shorter so I'm going to leave it, and I don't think it's like horribly obfuscating or anything.

File: test/inductor/test_compiled_optimizers.py (hunk 7 above)
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:91]: Yeah I can add a weight decay config and remove the foreach=True

File: test/inductor/test_compiled_optimizers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/inductor/test_compiled_optimizers.py:21]: hmm meh. not a fan, but not a blocker for this PR. This should either be abstracted or have some sort of explanation documented.

File: test/inductor/test_compiled_optimizers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/inductor/test_compiled_optimizers.py:34]: Yea, this is fine, was just intrigued

File: torch/optim/adam.py (hunk 5 above)
Comment (mlazos) [reply torch/optim/adam.py:435]: Hmm, no, the check is fine, this is a performance thing again (the check is slow). The step is on cuda if we're compiling. What I mean by the check doesn't matter anymore is that if say, the step was on cuda but some param is on CPU, the optimizer will still run and be fine. Inductor will decide whether cudagraphs should be run, if there are cpu tensors it just won't run cudagraphs.

File: test/inductor/test_compiled_optimizers.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        opt_eager = optim_cls(model_eager.parameters(), **kwargs)
+        opt_compiled = optim_cls(model_compiled.parameters(), **kwargs)
+        # run the patcher so that step has the expected structure
+        torch._dynamo.eval_frame.TorchPatcher.patch()
+
+        # unwrap step to avoid a deliberate graph break due to
+        # a limitation of functionalization/no_grad detection
+        # see the [Note on graph break] in optimizer.py
+        # This ignores the outer _use_grad_if_differentiable wrapper, which is fine for now
+        # as dynamo does not support differentiable optimizers anyway
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:51]:
```suggestion
        # This ignores the outer _use_grad_if_differentiable wrapper
        # and instead manually disables grad before calling step, which is fine
        # for now as dynamo does not support differentiable optimizers anyway
```

File: torch/optim/optimizer.py [hunk 11]
```diff
@@ -252,9 +252,9 @@ def __repr__(self):
 
     # Currently needed by Adam and AdamW
     def _cuda_graph_capture_health_check(self):
-        # If we are compiling, we take the capturable path automatically
-        # One caveat here is that if we are compiling, we *permit* step/param tensors to be on CPU
-        # so we do not explicitly enable the capturable flag. Inductor will decide whether cudagraphs
+        # If we are compiling, we take the capturable path automatically by setting the flag
+        # to true during tracing
+        # Inductor will decide whether cudagraphs
```
Comment (janeyx99) [inline torch/optim/optimizer.py:257]:
```suggestion
        # Note [torch.compile x capturable]
        # If we are compiling, we try to take the capturable path automatically by
        # setting the flag to True during tracing. Due to this, we skip all the checks
        # normally required for determining whether we can use CUDA graphs and
        # shunt the responsibility to torch.inductor. This saves time during tracing
        # since the checks are slow without sacrificing UX since inductor will warn
        # later if CUDA graphs cannot be enabled, e.g.,
        # https://github.com/pytorch/pytorch/blob/d3ba8901d8640eb16f88b2bfef9df7fa383d4b47/torch/_inductor/compile_fx.py#L390.
        # Thus, when compiling, inductor will determine if cudagraphs
```

File: torch/optim/adamw.py [hunk 12]
```diff
@@ -479,7 +481,9 @@ def _multi_tensor_adamw(
     if len(params) == 0:
         return
 
-    if capturable:
+    # If we are compiling, the compiler will determine if cudagraphs
+    # is applicable
```
Comment (janeyx99) [inline torch/optim/adamw.py:485]: Now all of these could be replaced
```suggestion
    # If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
```

**Review by janeyx99 — APPROVED** [review]
🚀 ⭐️

**Comment by eellison** [comment]
I think this is responsible for a 50s increase in TIMM dynamic shapes compilation time [131s → 180s](https://hud.pytorch.org/benchmark/timm_models/inductor_dynamic?startTime=Fri,%2030%20Jun%202023%2000:42:01%20GMT&stopTime=Fri,%2007%20Jul%202023%2000:42:01%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=main&lCommit=6bfd507c15f2d26212d3e2b9e581d9525bfd37d1&rBranch=main&rCommit=7bc181d374a54db4b1556d263be965f6a848eecb)
